token.go

```diff
@@ -3,65 +3,152 @@ package wallabago
 import (
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
+	"io"
 	"net/http"
 	"net/url"
 	"strings"
+	"time"
 )
 
-var token Token
+var token *Token
 
-// Token represents the object being returned from the oauth process at the API containing the access token, expire time, type of token, scope and a refresh token
+func setToken(newToken *Token) {
+	token = newToken
+}
+
+// Token represents the object being returned from the oauth process at the API
+// containing the access token, expire time (after converting it from the
+// number of seconds the token is valid to the point in time where it will
+// expires), type of token, scope and a refresh token
 type Token struct {
+	AccessToken    string
+	ExpirationTime time.Time
+	TokenType      string
+	Scope          string
+	RefreshToken   string
+}
+
+type tokenResponse struct {
 	AccessToken  string `json:"access_token"`
 	ExpiresIn    int    `json:"expires_in"`
 	TokenType    string `json:"token_type"`
-	Scope        string
+	Scope        string `json:"scope"`
 	RefreshToken string `json:"refresh_token"`
 }
 
+// parseTokenResponse consumes a stream (typically a http.Response.Body) that
+// is expected to contain JSON that unmarshals into a tokenResponse struct
+func parseTokenResponse(reader io.ReadCloser) (*tokenResponse, error) {
+	defer reader.Close()
+
+	var tokenResponse tokenResponse
+	err := json.NewDecoder(reader).Decode(&tokenResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tokenResponse, nil
+}
+
+// responseToToken converts a tokenResponse into a Token, computing the point
+// in time where the token will expire using the ExpiresIn field in
+// tokenResponse and the moment in which the function is called
+func responseToToken(tokenResponse *tokenResponse) *Token {
+	expiresIn := time.Duration(tokenResponse.ExpiresIn) * time.Second
+	expirationTime := time.Now().Add(expiresIn)
+
+	return &Token{
+		AccessToken:    tokenResponse.AccessToken,
+		ExpirationTime: expirationTime,
+		TokenType:      tokenResponse.TokenType,
+		Scope:          tokenResponse.Scope,
+		RefreshToken:   tokenResponse.RefreshToken,
+	}
+}
+
 // getToken will use the credentials set in the configuration to
 // request an access token from the wallabag API
-func getToken() (Token, error) {
-	var token Token
+func getToken() (*tokenResponse, error) {
 	tokenURL := Config.WallabagURL + "/oauth/v2/token"
 	resp, err := http.PostForm(tokenURL,
-		url.Values{"grant_type": {"password"},
+		url.Values{
+			"grant_type":    {"password"},
 			"client_id":     {Config.ClientID},
 			"client_secret": {Config.ClientSecret},
 			"username":      {Config.UserName},
 			"password":      {Config.UserPassword},
 		})
 	if err != nil {
-		return token, err
+		return nil, err
 	}
-	defer resp.Body.Close()
```

Comment:
Isn't it necessary to Close() anymore? I see that we use it in the return. Wouldn't it be better to get the result of parseTokenResponse(resp.Body), save it in a var and close the resp.Body and finally return the saved result?

Reply:
I embedded the Close inside the `parseTokenResponse`, but it makes sense to do it in the owner of the resp.Body object. Check my latest commit.